src/pip/_internal/resolution/resolvelib/found_candidates.py

```diff
@@ -0,0 +1,122 @@
+
+    def __next__(self):
+        # type: () -> Candidate
+        if self._installed and self._installed.version not in self._returned:
+            self._returned.add(self._installed.version)
+            return self._installed
+        if self._others is None:
+            self._others = self._get_others()
+        cand = next(self._others)
+        while cand.version in self._returned:
+            cand = next(self._others)
+        self._returned.add(cand.version)
+        return cand
+
+    next = __next__  # XXX: Python 2.
+
+
+class _InstalledReplacesCandidatesIterator(collections_abc.Iterator):
+    """Iterator for ``FoundCandidates``.
+
+    This iterator is used when the resolver prefers to upgrade an
+    already-installed package. Candidates from index are returned in their
+    normal ordering, except replaced when the version is already installed.
+    """
+    def __init__(
+        self,
+        get_others,  # type: Callable[[], Iterator[Candidate]]
+        installed,  # type: Optional[Candidate]
+    ):
+        self._installed = installed
+        self._get_others = get_others
+        self._others = None  # type: Optional[Iterator[Candidate]]
+        self._returned = set()  # type: Set[_BaseVersion]
+
+    def __next__(self):
+        # type: () -> Candidate
+        if self._others is None:
+            self._others = self._get_others()
+        cand = next(self._others)
+        while cand.version in self._returned:
+            cand = next(self._others)
+        if self._installed and cand.version == self._installed.version:
+            cand = self._installed
+        self._returned.add(cand.version)
+        return cand
+
+    next = __next__  # XXX: Python 2.
+
+
+class FoundCandidates(collections_abc.Sequence):
+    """A lazy sequence to provide candidates to the resolver.
+
+    The intended usage is to return this from `find_matches()` so the resolver
+    can iterate through the sequence multiple times, but only access the index
+    page when remote packages are actually needed. This improve performances
+    when suitable candidates are already installed on disk.
+    """
+    def __init__(
+        self,
+        get_others,  # type: Callable[[], Iterator[Candidate]]
+        installed,  # type: Optional[Candidate]
+        prefers_installed,  # type: bool
+    ):
+        self._get_others = get_others
+        self._installed = installed
+        self._prefers_installed = prefers_installed
+
+    def __getitem__(self, index):
+        # type: (int) -> Candidate
+        for i, value in enumerate(self):
+            if index == i:
+                return value
+        raise IndexError(index)
+
+    def __iter__(self):
+        # type: () -> Iterator[Candidate]
+        if self._prefers_installed:
+            klass = _InstalledFirstCandidatesIterator
+        else:
+            klass = _InstalledReplacesCandidatesIterator
+        return klass(self._get_others, self._installed)
```

Comment:
This separation is really nice, and is why I like your implementation more than mine. :)